### ISSUE TITLE
Normalisation methods

### ISSUE THREAD

**Issue by distracteddreamer** [distracteddreamer/mini-batch-learning#1]
### Batchnorm
**Description**
- Whiten the outputs of a layer before passing it through non-linearity
- Normalise each feature or filter (for conv nets) independently
- For a fully connected/dense layer with M examples and N dimensions find the batchwise mean and variance across all the M examples for each of the N dimensions, normalising each independently
- For a conv layer with M examples of size H x W x F, take the mean across all the M  H x W feature maps for each of the F filters 

**How it works**
- The original authors claimed it worked because it reduced internal covariate shift
- Internal covariate shift means the change in distributions of the outputs of layers as you go down a neural network
- For example if you have small variance for the distribution values of the inputs of a layer, this expands to a large range of values for the outputs. 
- By this reasoning it is not enough only to normalise the inputs for a deep network because as you go down the network the effect disappears. 
- Even if weights are well-initialised, this is does not prevent the variation in values of the inputs from affecting the outputs.
- The idea is therefore to whiten the inputs of every _subnetwork_

**When to normalise**
- We normalise the output of the dense or convolutional layer i.e.  u = Wx + b or u = conv(w, x) + b.
- Since x is generally the output of a non-linearity it is expected to have an unstable distribution, one that changes over training.
- On the other hand they claim that Wx + b is 'more likely to have a symmetric, non-sparse distribution, that is "more Gaussian"' so that normalising it makes more sense as it will lead to a stable distribution of inputs to the next layer

**Adding a linear transform**
- Changing the distribution of the inputs modifies what the network represents so we introduce a linear transform parameterised by two parameters gamma and beta (a pair for each feature/filter).
- The idea is that this will allow the network to learn parameters such as to recover the original activations if that were optimal.
